### DISCUSSION TITLE
Installing Infra on kind

### DISCUSSION THREAD

**Discussion by pdevine** [infrahq/infra#1993] [General]
Hey everyone! I was going to post this guide to installing on kind in the repo, but thought it would make more sense to kick it off here in the discussion forum for now to get more feedback.

# Installing Infra on kind

This guide will show you how to set up [kind](https://kind.sigs.k8s.io/) and Infra on Linux using the Metallb load balancer. Docker Desktop for macOS and Windows both use a virtual machine which runs Linux, and does not directly expose docker's network to the host. Consider using Docker's built-in kubernetes implementation instead of kind if you are using Docker Desktop.

This guide assumes that you have already [installed kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl-linux/) and [docker](https://docs.docker.com/engine/install/).

### 1. Download the latest version of kind

You'll first need to download kind. We'll use one of the pre-compiled binaries, however, you can also follow the [kind quickstart guide](https://kind.sigs.k8s.io/docs/user/quick-start/) for alternate methods.

We're assuming that you are using an Intel 64 bit instance, but there are other releases available [here](https://github.com/kubernetes-sigs/kind/releases/).

```
curl -Lo ./kind https://kind.sigs.k8s.io/dl/v0.13.0/kind-linux-amd64
chmod +x ./kind
mv ./kind /some-dir-in-your-PATH/kind
```

### 2. Create a cluster

Once you have the kind binary configured in your path, create your kubernetes cluster using the command:

```
kind create cluster
```

### 3. Create a load balancer

Now that your kubernetes cluster is up, we'll install the Metallb load balancer.

```
kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/v0.12.1/manifests/namespace.yaml
kubectl apply -f https://raw.githubusercontent.com/metallb/metallb/v0.12.1/manifests/metallb.yaml
```

You can verify that the load balancer is running with the command:

```
kubectl get pods -n metallb-system
```

### 4. Determine the IP address pool for Docker

We still need to set up the IP address pool which Metallb will use. Since we're using docker, we need to look at the docker networking stack to determine which IP addresses we can use.

```
docker network inspect -f '{{ (index .IPAM.Config 0).Subnet }}' kind
```

This will print out something similar to:

```
172.19.0.0/16
```

In this case our docker network is set to use `172.19.0.0/16`, however it may be different on your Linux host. We'll configure the load balancer to use IPs in the range `172.19.255.200-172.19.255.250` since it's unlikely for Docker to assign any of those IPs to containers. 

Next, create a ConfigMap called `metallb-configmap.yaml` with IPs in docker's network range. It should look something like:

```
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: metallb-system
  name: config
data:
  config: |
    address-pools:
    - name: default
      protocol: layer2
      addresses:
      - 172.19.255.200-172.19.255.250
```

Once you're finished, apply the ConfigMap with the command:

`kubectl apply -f metallb-configmap.yaml`

### 5. Follow the rest of the Quickstart guide

The rest of the steps to get Infra connected should be the same as a standard Kubernetes installation.  Open up the [Quickstart guide](https://infrahq.com/docs/getting-started/quickstart) and you can complete installing Infra from the section "Install Infra CLI".

**Comment by pdevine** [comment]
After writing this up, I realized that there is a way for making this work inside Docker Desktop, however, you'll need to first set up [Docker Mac Net Connect](https://github.com/chipmk/docker-mac-net-connect). The problem is that by default the Docker Linux VM won't route anything onto the bridge network setup by `kind`.

To get this to work, as ***Step 0*** (i.e. before setting up anything else), do the following:

```
# Install via Homebrew
$ brew install chipmk/tap/docker-mac-net-connect

# Run the service and register it to launch at boot
$ sudo brew services start chipmk/tap/docker-mac-net-connect
```

This will automatically run in the background and set up Wireguard in the Linux VM and will automatically create new routes when new networks are created in Docker. To state the obvious, you don't want to do anything like this in production, but it works reasonably well if you want to use *kind* instead of the built in Docker Desktop kubernetes.

**Reply by technovangelist** [reply]
> To state the obvious, you don't want to do anything like this in production, but it works reasonably well if you want to use kind instead of the built in Docker Desktop kubernetes

How about if you want to use **kind** *as well as* **docker desktop k8s**???

**Reply by pdevine** [reply]
I'm not 100% sure, but I think that will work just fine. Docker Desktop's load balancer will be routed separately from Wireguard+kind. You should then be able to just switch contexts to go between the two installations.
